File: src/ipc/codebook.ts
import { invoke } from "@tauri-apps/api/core";

export type Cluster = {
  id: number;
  name: string;
  description: string | null;
  color: string | null;
  orderIndex: number;
};

export type Category = {
  id: number;
  clusterId: number | null;
  name: string;
  description: string | null;
  color: string | null;
  orderIndex: number;
};

export type Tag = {
  id: number;
  categoryId: number | null;
  name: string;
  description: string | null;
  color: string | null;
  orderIndex: number;
};

export type TagNode = {
  id: number;
  categoryId: number | null;
  name: string;
  description: string | null;
  color: string | null;
  orderIndex: number;
  count: number;
};

export type CategoryNode = {
  id: number;
  clusterId: number | null;
  name: string;
  description: string | null;
  color: string | null;
  orderIndex: number;
  count: number;
  tags: TagNode[];
};

export type ClusterNode = {
  id: number;
  name: string;
  description: string | null;
  color: string | null;
  orderIndex: number;
  count: number;
  categories: CategoryNode[];
};

export type CodebookTree = {
  clusters: ClusterNode[];
  standaloneCategories: CategoryNode[];
  standaloneTags: TagNode[];
};

type RawCluster = {
  id: number;
  name: string;
  description: string | null;
  color: string | null;
  order_index: number;
};
type RawCategory = RawCluster & { cluster_id: number | null };
type RawTag = RawCluster & { category_id: number | null };
type RawTagNode = RawTag & { count: number };
type RawCategoryNode = RawCategory & { count: number; tags: RawTagNode[] };
type RawClusterNode = RawCluster & {
  count: number;
  categories: RawCategoryNode[];
};
type RawCodebookTree = {
  clusters: RawClusterNode[];
  standalone_categories: RawCategoryNode[];
  standalone_tags: RawTagNode[];
};

const tagFromRaw = (r: RawTag): Tag => ({
  id: r.id,
  categoryId: r.category_id,
  name: r.name,
  description: r.description,
  color: r.color,
  orderIndex: r.order_index,
});

const tagNodeFromRaw = (r: RawTagNode): TagNode => ({
  ...tagFromRaw(r),
  count: r.count,
});

const categoryFromRaw = (r: RawCategory): Category => ({
  id: r.id,
  clusterId: r.cluster_id,
  name: r.name,
  description: r.description,
  color: r.color,
  orderIndex: r.order_index,
});

const categoryNodeFromRaw = (r: RawCategoryNode): CategoryNode => ({
  ...categoryFromRaw(r),
  count: r.count,
  tags: r.tags.map(tagNodeFromRaw),
});

const clusterFromRaw = (r: RawCluster): Cluster => ({
  id: r.id,
  name: r.name,
  description: r.description,
  color: r.color,
  orderIndex: r.order_index,
});

const clusterNodeFromRaw = (r: RawClusterNode): ClusterNode => ({
  ...clusterFromRaw(r),
  count: r.count,
  categories: r.categories.map(categoryNodeFromRaw),
});

const treeFromRaw = (r: RawCodebookTree): CodebookTree => ({
  clusters: r.clusters.map(clusterNodeFromRaw),
  standaloneCategories: r.standalone_categories.map(categoryNodeFromRaw),
  standaloneTags: r.standalone_tags.map(tagNodeFromRaw),
});

export const codebookTree = async (): Promise<CodebookTree> =>
  treeFromRaw(await invoke<RawCodebookTree>("codebook_tree"));

export const clusterCreate = async (
  name: string,
  description?: string | null,
  color?: string | null,
): Promise<Cluster> =>
  clusterFromRaw(
    await invoke<RawCluster>("cluster_create", {
      name,
      description: description ?? null,
      color: color ?? null,
    }),
  );
export const clusterRename = (id: number, name: string): Promise<void> =>
  invoke("cluster_rename", { id, name });
export const clusterSetDescription = (
  id: number,
  description: string | null,
): Promise<void> => invoke("cluster_set_description", { id, description });
export const clusterSetColor = (
  id: number,
  color: string | null,
): Promise<void> => invoke("cluster_set_color", { id, color });
export const clusterDelete = (id: number): Promise<void> =>
  invoke("cluster_delete", { id });
export const clusterReorder = (ids: number[]): Promise<void> =>
  invoke("cluster_reorder", { ids });

export const categoryCreate = async (
  clusterId: number | null,
  name: string,
  description?: string | null,
  color?: string | null,
): Promise<Category> =>
  categoryFromRaw(
    await invoke<RawCategory>("category_create", {
      clusterId,
      name,
      description: description ?? null,
      color: color ?? null,
    }),
  );
export const categoryRename = (id: number, name: string): Promise<void> =>
  invoke("category_rename", { id, name });
export const categorySetDescription = (
  id: number,
  description: string | null,
): Promise<void> => invoke("category_set_description", { id, description });
export const categorySetColor = (
  id: number,
  color: string | null,
): Promise<void> => invoke("category_set_color", { id, color });
export const categoryDelete = (id: number): Promise<void> =>
  invoke("category_delete", { id });
export const categoryReorder = (
  clusterId: number | null,
  ids: number[],
): Promise<void> => invoke("category_reorder", { clusterId, ids });
export const categoryMoveToCluster = (
  id: number,
  newClusterId: number | null,
): Promise<void> => invoke("category_move_to_cluster", { id, newClusterId });

export const tagCreate = async (
  categoryId: number | null,
  name: string,
  description?: string | null,
  color?: string | null,
): Promise<Tag> =>
  tagFromRaw(
    await invoke<RawTag>("tag_create", {
      categoryId,
      name,
      description: description ?? null,
      color: color ?? null,
    }),
  );
export const tagRename = (id: number, name: string): Promise<void> =>
  invoke("tag_rename", { id, name });
export const tagSetDescription = (
  id: number,
  description: string | null,
): Promise<void> => invoke("tag_set_description", { id, description });
export const tagSetColor = (id: number, color: string | null): Promise<void> =>
  invoke("tag_set_color", { id, color });
export const tagDelete = (id: number): Promise<void> =>
  invoke("tag_delete", { id });
export const tagReorder = (
  categoryId: number,
  ids: number[],
): Promise<void> => invoke("tag_reorder", { categoryId, ids });
export const tagMoveToCategory = (
  id: number,
  newCategoryId: number | null,
): Promise<void> => invoke("tag_move_to_category", { id, newCategoryId });

File: src-tauri/src/commands/codebook.rs
use crate::commands::util::project_conn;
use crate::db::queries::{category, cluster, stats, tag};
use crate::domain::codebook::{CategoryNode, ClusterNode, CodebookTree, TagNode};
use crate::error::AppResult;
use rusqlite::Connection;
use std::collections::HashMap;

pub fn build_tree(conn: &Connection) -> AppResult<CodebookTree> {
    let counts = stats::codebook_counts(conn)?;
    let clusters = cluster::list(conn)?;
    let categories = category::list_all(conn)?;
    let tags = tag::list_all(conn)?;

    let mut by_cat: HashMap<i64, Vec<TagNode>> = HashMap::new();
    let mut standalone_tags: Vec<TagNode> = Vec::new();
    for t in tags {
        let count = counts.by_tag.get(&t.id).copied().unwrap_or(0);
        let node = TagNode {
            id: t.id,
            category_id: t.category_id,
            name: t.name,
            description: t.description,
            color: t.color,
            order_index: t.order_index,
            count,
        };
        match t.category_id {
            Some(cid) => by_cat.entry(cid).or_default().push(node),
            None => standalone_tags.push(node),
        }
    }

    let mut by_cluster: HashMap<i64, Vec<CategoryNode>> = HashMap::new();
    let mut standalone_categories: Vec<CategoryNode> = Vec::new();
    for c in categories {
        let count = counts.by_category.get(&c.id).copied().unwrap_or(0);
        let node = CategoryNode {
            id: c.id,
            cluster_id: c.cluster_id,
            name: c.name,
            description: c.description,
            color: c.color,
            order_index: c.order_index,
            count,
            tags: by_cat.remove(&c.id).unwrap_or_default(),
        };
        match c.cluster_id {
            Some(cluster_id) => by_cluster.entry(cluster_id).or_default().push(node),
            None => standalone_categories.push(node),
        }
    }

    let mut out_clusters = Vec::new();
    for c in clusters {
        let count = counts.by_cluster.get(&c.id).copied().unwrap_or(0);
        let cats = by_cluster.remove(&c.id).unwrap_or_default();
        out_clusters.push(ClusterNode {
            id: c.id,
            name: c.name,
            description: c.description,
            color: c.color,
            order_index: c.order_index,
            count,
            categories: cats,
        });
    }

    Ok(CodebookTree {
        clusters: out_clusters,
        standalone_categories,
        standalone_tags,
    })
}

#[tauri::command]
pub async fn codebook_tree(app: tauri::AppHandle) -> AppResult<CodebookTree> {
    let conn = project_conn(&app)?;
    build_tree(&conn)
}

// -- cluster commands -------------------------------------------------------

#[tauri::command]
pub async fn cluster_create(
    app: tauri::AppHandle,
    name: String,
    description: Option<String>,
    color: Option<String>,
) -> AppResult<cluster::Cluster> {
    let conn = project_conn(&app)?;
    cluster::create(&conn, &name, description.as_deref(), color.as_deref())
}

#[tauri::command]
pub async fn cluster_rename(app: tauri::AppHandle, id: i64, name: String) -> AppResult<()> {
    let conn = project_conn(&app)?;
    cluster::rename(&conn, id, &name)
}

#[tauri::command]
pub async fn cluster_set_description(
    app: tauri::AppHandle,
    id: i64,
    description: Option<String>,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    cluster::set_description(&conn, id, description.as_deref())
}

#[tauri::command]
pub async fn cluster_set_color(
    app: tauri::AppHandle,
    id: i64,
    color: Option<String>,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    cluster::set_color(&conn, id, color.as_deref())
}

#[tauri::command]
pub async fn cluster_delete(app: tauri::AppHandle, id: i64) -> AppResult<()> {
    let conn = project_conn(&app)?;
    cluster::delete(&conn, id)
}

#[tauri::command]
pub async fn cluster_reorder(app: tauri::AppHandle, ids: Vec<i64>) -> AppResult<()> {
    let mut conn = project_conn(&app)?;
    cluster::reorder(&mut conn, &ids)
}

// -- category commands ------------------------------------------------------

#[tauri::command]
pub async fn category_create(
    app: tauri::AppHandle,
    cluster_id: Option<i64>,
    name: String,
    description: Option<String>,
    color: Option<String>,
) -> AppResult<category::Category> {
    let conn = project_conn(&app)?;
    category::create(
        &conn,
        cluster_id,
        &name,
        description.as_deref(),
        color.as_deref(),
    )
}

#[tauri::command]
pub async fn category_rename(app: tauri::AppHandle, id: i64, name: String) -> AppResult<()> {
    let conn = project_conn(&app)?;
    category::rename(&conn, id, &name)
}

#[tauri::command]
pub async fn category_set_description(
    app: tauri::AppHandle,
    id: i64,
    description: Option<String>,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    category::set_description(&conn, id, description.as_deref())
}

#[tauri::command]
pub async fn category_set_color(
    app: tauri::AppHandle,
    id: i64,
    color: Option<String>,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    category::set_color(&conn, id, color.as_deref())
}

#[tauri::command]
pub async fn category_delete(app: tauri::AppHandle, id: i64) -> AppResult<()> {
    let conn = project_conn(&app)?;
    category::delete(&conn, id)
}

#[tauri::command]
pub async fn category_reorder(
    app: tauri::AppHandle,
    cluster_id: Option<i64>,
    ids: Vec<i64>,
) -> AppResult<()> {
    let mut conn = project_conn(&app)?;
    category::reorder(&mut conn, cluster_id, &ids)
}

#[tauri::command]
pub async fn category_move_to_cluster(
    app: tauri::AppHandle,
    id: i64,
    new_cluster_id: Option<i64>,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    category::move_to_cluster(&conn, id, new_cluster_id)
}

// -- tag commands -----------------------------------------------------------

#[tauri::command]
pub async fn tag_create(
    app: tauri::AppHandle,
    category_id: Option<i64>,
    name: String,
    description: Option<String>,
    color: Option<String>,
) -> AppResult<tag::Tag> {
    let conn = project_conn(&app)?;
    tag::create(
        &conn,
        category_id,
        &name,
        description.as_deref(),
        color.as_deref(),
    )
}

#[tauri::command]
pub async fn tag_rename(app: tauri::AppHandle, id: i64, name: String) -> AppResult<()> {
    let conn = project_conn(&app)?;
    tag::rename(&conn, id, &name)
}

#[tauri::command]
pub async fn tag_set_description(
    app: tauri::AppHandle,
    id: i64,
    description: Option<String>,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    tag::set_description(&conn, id, description.as_deref())
}

#[tauri::command]
pub async fn tag_set_color(
    app: tauri::AppHandle,
    id: i64,
    color: Option<String>,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    tag::set_color(&conn, id, color.as_deref())
}

#[tauri::command]
pub async fn tag_delete(app: tauri::AppHandle, id: i64) -> AppResult<()> {
    let conn = project_conn(&app)?;
    tag::delete(&conn, id)
}

#[tauri::command]
pub async fn tag_reorder(
    app: tauri::AppHandle,
    category_id: i64,
    ids: Vec<i64>,
) -> AppResult<()> {
    let mut conn = project_conn(&app)?;
    tag::reorder(&mut conn, category_id, &ids)
}

#[tauri::command]
pub async fn tag_move_to_category(
    app: tauri::AppHandle,
    id: i64,
    new_category_id: Option<i64>,
) -> AppResult<()> {
    let conn = project_conn(&app)?;
    tag::move_to_category(&conn, id, new_category_id)
}
